Reuse balance transaction in construct transaction

paweljakubas opened the pull request:
<!--
Detail in a few bullet points the work accomplished in this PR.

Before you submit, don't forget to:

* Make sure the GitHub PR fields are correct:
   ✓ Set a good Title for your PR.
   ✓ Assign yourself to the PR.
   ✓ Assign one or more reviewer(s).
   ✓ Link to a Jira issue, and/or other GitHub issues or PRs.
   ✓ In the PR description delete any empty sections
     and all text commented in <!--, so that this text does not appear
     in merge commit messages.

* Don't waste reviewers' time:
   ✓ If it's a draft, select the Create Draft PR option.
   ✓ Self-review your changes to make sure nothing unexpected slipped through.

* Try to make your intent clear:
   ✓ Write a good Description that explains what this PR is meant to do.
   ✓ Jira will detect and link to this PR once created, but you can also
     link this PR in the description of the corresponding Jira ticket.
   ✓ Highlight what Testing you have done.
   ✓ Acknowledge any changes required to the Documentation.
-->


- [x] Added PreSelection along side SelectionOf
- [x] Enabled mkUnsignedTx to use either PreSelection or SelectionOf
- [x] constructTransaction produces "no-inputs/no-fee" sealed tx using mkUnsignedTx
- [x] constructTransaction uses balanceTransaction to add inputs and fees
- [x] coin selection and fee are retrieved from decodeTransaction
- [x] checking all integration tests
- [x] adding dummyTxIn as node does not allow producing no-inputs txs at the moment  

### Comments

<!-- Additional comments, links, or screenshots to attach, if any. -->

### Issue Number
adp-2257

<!-- Reference the Jira/GitHub issue that this PR relates to, and which requirements it tackles.
  Note: Jira issues of the form ADP- will be auto-linked. -->

paweljakubas committed:
generalize balanceTransaction

paweljakubas committed:
add preselection

paweljakubas committed:
use PreSelection in mkUnsignedTransaction

paweljakubas committed:
refactor: constructTransaction

paweljakubas committed:
introduce dummyInput

Anviking committed:
Make contents of dummyInput stand out more

Anviking committed:
Remove dummy input immediately after calling makeTransactionBody

paweljakubas committed:
adding minting script witness to estimateNumberOfWitnesses

paweljakubas committed:
be precise about number of minting scripts

paweljakubas committed:
update test of withdrawal from self, 0 rewards

jonathanknowles committed:
Fix imports.

jonathanknowles committed:
Miscellaneous formatting fixes.

jonathanknowles committed:
Make inner functions of `constructTransaction` more concise.

Use `LambdaCase` instead of repeating patterns.
Use `Coin.fromQuantity` instead of `fromIntegral`.

jonathanknowles committed:
Factor out common build command within definition of `inputWits`.

jonathanknowles committed:
Remove redundant `IsCardanoEra` constraint.

jonathanknowles committed:
Match all cases in function `removeDummyInput`.

This removes the incomplete pattern match warnings.

jonathanknowles committed:
Add reminder to remove the `dummyInput` hack.

jonathanknowles committed:
Simplify type of `Cardano.Wallet.constructTransaction`.

The function `Cardano.Wallet.constructTransaction` is only ever called
with a `Left PreSelection`, and never called with a `Right Selection`.

Therefore, we can simplify the selection argument type from
`Either PreSelection Selection` to just `PreSelection`.

jonathanknowles commented on lib/wallet/src/Cardano/Wallet.hs:
```diff
@@ -2534,7 +2534,7 @@ constructTransaction
     -> Cardano.AnyCardanoEra
     -> TransactionCtx
-    -> SelectionOf TxOut
+    -> Either PreSelection (SelectionOf TxOut)
```
Hi @paweljakubas 

I'm a bit confused about why this argument needs to be of type `Either PreSelection Selection`.

Is there a good reason for this? (Perhaps preparing for the future?)

AFAICT, this function is only ever called with `Left PreSelection`, so we could just simplify the argument type to `PreSelection`.

See the following commit, which still compiles: https://github.com/input-output-hk/cardano-wallet/pull/3567/commits/262b7b939ae9d64c242f50410b1b3209858397f2.

paweljakubas replied:
hi @jonathanknowles indeed. `PreSelection` was introduced as the intention was to construct tx WITHOUT inputs and `SelectionOf` has `NonEmpty` of inputs. But when I proceeded then it turn out that node does not accept constructing tx with empty inputs. Hence this `dummyInput` hack. But, node intends to lift this restriction in coming version of node, so I intended to stick to differentiating `PreSelection` and `SelectionOf`.

Anviking commented on lib/wallet/src/Cardano/Wallet/CoinSelection.hs:
```diff
@@ -384,6 +385,27 @@ toExternalSelectionSkeleton Internal.SelectionSkeleton {..} =
 
+-- | Represents a unbalanced selection.
+--
+data PreSelection = PreSelection
```
Could this be moved? E.g. to alongside `TransactionCtx`?

paweljakubas replied:
yeah, sounds like good idea ...wonder only if it will not major ripple effects and affect other calls (other than constructTransaction)

jonathanknowles replied:
@paweljakubas @Anviking I've moved this in https://github.com/input-output-hk/cardano-wallet/pull/3553/commits/f154f16a2633879d74fc33cd64e44a78c55a74f0 so that it is next to `TransactionCtx`.

jonathanknowles committed:
Move type `PreSelection` to `Cardano.Wallet.Transaction`.

This helper type is only used by the `constructTransaction` and
`mkUnsignedTx` functions, and not used by anything within the
`CoinSelection` module hierarchy.

Anviking committed:
Tweak retrospective computation of change & outs

Without this change integration tests would fail e.g. where:

```pseudo-code
constructTx
  paymentTo addr1 (0 ada)

  |
  |
  V

Tx
  output: addr1 (1 ada) -- min ada, set automatically
  output: addr2 (9 ada) -- change
```

as we'd believe both outputs in the resuling tx were change -- neither
of them exists in the original output list.

If there are 'n' outputs in the original unbalanced tx, we will with
this commit instead judge that the first 'n' are "outputs" and that the
ones after that are all change. While this makes assumptions about how
balanceTx appends outputs, it is more importantly ressilient to outputs
being modified, like with the `increaseZeroAdaOutputs` feature.

Unisay committed:
refactor: explicit parameters for the Wallet.constructTransaction

Unisay committed:
chore: code formatting

Unisay committed:
chore: improve comment on PreSelection data type

Unisay committed:
fix: removeDummyInput does nothing for non-babbage txs.

Unisay committed:
feat: remove a dummy input from a Mary tx body

Anviking commented on lib/wallet/src/Cardano/Wallet/Shelley/Transaction.hs:
```diff
@@ -2450,7 +2450,20 @@ removeDummyInput txBody =
                 ShelleyBasedEraShelley -> txBody
                 ShelleyBasedEraAllegra -> txBody
                 ShelleyBasedEraMary -> txBody
```
@Unisay could we please revert 0fe8b4f8331f82e322c389c6dd27719d93938b1f and preserve the `bailOut` behaviour for Mary and earlier, rather than silently not removing the dummyInput?

Unisay replied:
Done

Unisay committed:
Safer dummy input removal

Anviking commented on lib/wallet/src/Cardano/Wallet/Shelley/Transaction.hs:
```diff
@@ -1056,16 +1063,32 @@ estimateNumberOfWitnesses (Cardano.TxBody txbodycontent) =
         txUpdateProposal' +
-        txCerts
+        txCerts +
+        txMintingScripts
```
This is wrong, but I suspect it's not a regression, and is fine for now.

We should estimate the upper bound of key witnesses required for all scripts (and ideally never counting the same key twice).

This would be needed if we were to _pre-select_ inputs from a shared wallet, or if we have a more complicated minting script (e.g. needing 2 signers).

It would probably be good to re-use the logic in: [`estimateMaxWitnessRequiredPerInput`](https://github.com/input-output-hk/cardano-wallet/blob/157a6d5f977f4600373596b7cfa9700138e8e140/lib/wallet/src/Cardano/Wallet/Primitive/AddressDiscovery/Shared.hs#L772-L784)

Anviking replied:
No this might have been a regression. 

This should be the old behaviour: https://github.com/input-output-hk/cardano-wallet/blob/2dfeac9e660f75be262877b77db5da041ad1aa4d/lib/wallet/src/Cardano/Wallet/Shelley/Transaction.hs#L1772-L1773 Which looks correct.

Anviking commented on lib/wallet/src/Cardano/Wallet/Shelley/Transaction.hs:
```diff
@@ -2478,6 +2478,8 @@ removeDummyInput txBody =
                         aux
                         val
+  where
+    bailOut = error "removeDummyInput only supported for the Babbage era"
```
Nitpick:
```suggestion
    bailOut = error "removeDummyInput only supported for the Alonzo and Babbage era"
```

Anviking commented on lib/wallet/api/http/Cardano/Wallet/Api/Http/Shelley/Server.hs:
```diff
@@ -2440,68 +2424,134 @@ constructTransaction ctx genChange knownPools getPoolStatus (ApiT wid) body = do
+            W.constructTransaction @n @'CredFromKeyK
+                txLayer netLayer db wid era txCtx' PreSelection
+                    { outputs = outs <> mintingOuts
+                    , assetsToMint = fst $ txCtx' ^. #txAssetsToMint
+                    , assetsToBurn = fst $ txCtx' ^. #txAssetsToBurn
+                    , extraCoinSource = fromMaybe (Coin 0) refund
+                    , extraCoinSink = fromMaybe (Coin 0) deposit
```
🤔 All these aside from `outputs` could be `mempty` (mkUnsignedTx doesn't look at them & ran integration tests - it works), so it seems the `PreSelection` datatype could be replaced with just `[TxOut]`. (Or possibly integrated with `TransactionCtx` 🤷‍♂️ )

@Unisay I recall you asked about this and I might have mislead you: it's _these_ similarly named fields https://github.com/input-output-hk/cardano-wallet/blob/7e7a60b9a3f23b7b726157fb721f1c1dfb7805a2/lib/wallet/src/Cardano/Wallet/CoinSelection.hs#L280-L290 in `SelectionParams` that currently are unavoidable, but the ones in `PreSelection` seem redundant.

Anviking replied:
(Feel free to ignore for now and tackle when/how you see fit)

Unisay committed:
refactor: better structured validation logic in the constructTransaction
